src/components/Clipboard.react.js

```diff
@@ -0,0 +1,147 @@
+import React, {Component} from 'react'; // eslint-disable-line no-unused-vars
+import PropTypes from 'prop-types';
+import {FontAwesomeIcon} from '@fortawesome/react-fontawesome';
+import {faCopy, faCheckCircle} from '@fortawesome/free-regular-svg-icons';
+
+const clipboardAPI = navigator.clipboard;
+
+/**
+ * The Clipboard component copies text to the clipboard
+ */
+
+export default class Clipboard extends React.Component {
+    constructor(props) {
+        super(props);
+        this.copyToClipboard = this.copyToClipboard.bind(this);
+        this.copySuccess = this.copySuccess.bind(this);
+        this.getText = this.getText.bind(this);
+        this.stringifyId = this.stringifyId.bind(this);
+        this.state = {
+            copied: false,
+        };
+    }
+
+    // stringifies object ids used in pattern matching callbacks
+    stringifyId(id) {
+        if (typeof id !== 'object') {
+            return id;
+        }
+        const stringifyVal = v => (v && v.wild) || JSON.stringify(v);
+        const parts = Object.keys(id)
+            .sort()
+            .map(k => JSON.stringify(k) + ':' + stringifyVal(id[k]));
+        return '{' + parts.join(',') + '}';
+    }
+
+    copySuccess() {
+        this.props.setProps({
+            n_clicks: this.props.n_clicks + 1,
+        });
+
+        const showCopiedIcon = 1000;
+        this.setState({copied: true}, () => {
+            setTimeout(() => {
+                this.setState({copied: false});
+            }, showCopiedIcon);
+        });
+    }
+
+    getText() {
+        // get the inner text.  If none, use the content of the value param
+        const id = this.stringifyId(this.props.target_id);
+        const target = document.getElementById(id);
+        let text = target.innerText;
+        if (!text) {
+            text = target.value;
+            text = text === undefined ? null : text;
+        }
+        return text;
+    }
+
+    copyToClipboard() {
+        let text = this.props.text;
+        if (this.props.target_id) {
+            text = this.getText();
+        }
+        if (text) {
+            clipboardAPI.writeText(text).then(this.copySuccess, function() {
+                alert('copy error');
+            });
+        } else {
+            this.copySuccess();
```

Comment:
This is the "just use this icon as a button" case where you specified neither `props.text` nor `target_id`, right? But as it is we'll also get here if `props.text` or the target text is `''`.

I also don't think we want to promote methods like `pandas.DataFrame.to_clipboard` - that'll work fine when you're running the app locally, but once deployed it will fail. So seems to me this branch of calling `copySuccess` without actually having written anything to the clipboard is not a good idea. I'm not sure though what we _should_ do when there's nothing to copy - put a blank string in the clipboard and still show the check mark? Show an X mark instead (for the same short duration) and don't write anything to the clipboard? I guess I'd lean toward the latter.

Reply:
OK, I'll take out the "just use the icon as a button." 

It's pretty clear that the copy doesn't work when you click on the icon and nothing happens. 

And It's probably best to copy nothing to the clipboard in this case.  An empty string will clear out the current content.